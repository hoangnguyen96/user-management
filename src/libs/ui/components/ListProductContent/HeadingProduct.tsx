import { memo } from "react";
import { Stack, Typography } from "@mui/material";

const HeadingProduct = () => (
  <Stack
    flexDirection="row"
    justifyContent="flex-start"
    p="0 40px 14px"
    gap="16px"
  >
    <Typography variant="subtitle2" width={{ xs: "80px", lg: "140px" }}>
      Product Code
    </Typography>
    <Typography variant="subtitle2" width={{ xs: "110px", lg: "170px" }}>
      Name
    </Typography>
    <Typography variant="subtitle2" width={{ xs: "80px", lg: "140px" }}>
      Price
    </Typography>
    <Typography variant="subtitle2" width={{ xs: "60px", lg: "120px" }}>
      Quantity
    </Typography>
    <Typography variant="subtitle2" width={{ xs: "70px", lg: "120px" }}>
      Image
    </Typography>
    <Typography variant="subtitle2" ml="12px">
      Status
    </Typography>
  </Stack>
);

export default memo(HeadingProduct);
