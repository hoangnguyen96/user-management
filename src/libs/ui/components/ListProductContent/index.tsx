import {
  ChangeEvent,
  memo,
  useCallback,
  useEffect,
  useMemo,
  useState,
  useTransition,
} from "react";
import { Box } from "@mui/material";

// Constants
import { ROUTES, SORT_DATA_PRODUCT } from "@app/constants";

// Api
import {
  useCreateProduct,
  useGetListCustomers,
  useGetListProduct,
  useGetListProductOfUser,
} from "@app/api";

// Models
import { Product } from "@app/models";

// Utils
import {
  filterListProductByName,
  dividePaginationProduct,
  sortProductList,
} from "@app/utils";

// Components
import HeadingSearch from "../HeadingSearch";
import ListItemProduct from "./ListItemProduct";
import { useMatch } from "react-router-dom";
import ModalCreateProduct from "../Modal/ModalCreateProduct";

interface ListProductContentProps {
  isAdmin?: boolean;
  id: string;
}

const ListProductContent = ({
  isAdmin = false,
  id,
}: ListProductContentProps) => {
  const [isOpenModalCreateProduct, setIsOpenModalCreateProduct] =
    useState<boolean>(false);
  const [pagination, setPagination] = useState<number>(0);
  const [listPage, setListPage] = useState<Product[][]>([]);
  const [textSearch, setTextSearch] = useState<string>("");
  const [filterText, setFilterText] = useState<string>("");
  const [isPending, startTransition] = useTransition();
  const [selectionValue, setSelectionValue] = useState(
    SORT_DATA_PRODUCT[0].value
  );

  const isMatchProduct = !!useMatch(ROUTES.PRODUCT);

  const {
    isLoading: isLoadingListAll,
    refetch: refetchListAll,
    data: productListData,
    errorMessage: errorGetList,
  } = useGetListProduct();

  const {
    isLoading: isLoadingListOfUser,
    refetch: refetchListOfUser,
    data: productListOfUser,
    errorMessage: errorGetListOfUser,
  } = useGetListProductOfUser(id);

  const { data: listUser } = useGetListCustomers();

  const { isPending: isLoadingCreate, mutate: createProduct } =
    useCreateProduct();

  const productList = useMemo(() => {
    if (!isAdmin) {
      return productListOfUser;
    }

    return errorGetList || productListData.length <= 0 ? [] : productListData;
  }, [isAdmin, productListData, productListOfUser]);

  // Search Bar
  useEffect(() => {
    if (!isLoadingListAll) {
      if (filterText) {
        const listByName = filterListProductByName(productList, filterText);
        const filteredList = dividePaginationProduct(listByName);
        setListPage(filteredList);
      } else {
        const list = dividePaginationProduct(productList);
        setListPage(list);
      }
    }
  }, [filterText]);

  // Selection Bar
  useEffect(() => {
    if (!isLoadingListAll) {
      const sortedProductList = sortProductList(productList, selectionValue);
      if (sortedProductList.length > 0) {
        const list = dividePaginationProduct(sortedProductList);
        setListPage(list);
      }
    }
  }, [selectionValue]);

  // Reset list
  useEffect(() => {
    const listAll = dividePaginationProduct(productList);
    setListPage(listAll);
  }, [isLoadingListAll, isLoadingListOfUser]);

  const handleChangePagination = useCallback(
    (_event: ChangeEvent<unknown>, value: number) => {
      setPagination(value - 1);
    },
    []
  );

  const handleOnChangeText = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      setTextSearch(e.target.value);

      startTransition(() => {
        setFilterText(e.target.value);
      });
    },
    [setFilterText, startTransition]
  );

  const handleOpenModalCreateProduct = useCallback(() => {
    setIsOpenModalCreateProduct(true);
  }, [setIsOpenModalCreateProduct]);

  const handleCloseModalCreateProduct = useCallback(() => {
    setIsOpenModalCreateProduct(false);
  }, [setIsOpenModalCreateProduct]);

  const handleCreateProductSuccess = useCallback(() => {
    setIsOpenModalCreateProduct(false);
    refetchListAll();
  }, [refetchListAll, setIsOpenModalCreateProduct]);

  const handleCreateProductFailed = useCallback(
    () => alert("Create Product failed!"),
    []
  );

  const handleCreateProduct = useCallback((payload: Partial<Product>) => {
    createProduct(payload, {
      onSuccess: handleCreateProductSuccess,
      onError: handleCreateProductFailed,
    });
  }, []);

  const currentPageData = listPage[pagination] || [];

  return (
    <>
      <Box
        sx={{
          borderRadius: "24px",
          backgroundColor: "white",
          boxShadow: "1px 1px 2px #cccccc",
        }}
      >
        <HeadingSearch
          title="All Product"
          subTitle="Active Product"
          isProduct={isMatchProduct}
          textSearch={textSearch}
          list={SORT_DATA_PRODUCT}
          handleOnChangeText={handleOnChangeText}
          onOpenModalProduct={handleOpenModalCreateProduct}
          setSelectionValue={setSelectionValue}
          setTextSearch={setTextSearch}
        />

        {/* List */}
        <ListItemProduct
          isAdmin={isAdmin}
          isLoading={isLoadingListAll}
          isPending={isPending}
          userId={id}
          listCurrent={currentPageData}
          listPage={listPage}
          error={errorGetList || errorGetListOfUser}
          pagination={pagination}
          paginationList={productList}
          refetchList={refetchListAll || refetchListOfUser}
          onChangePagination={handleChangePagination}
        />
      </Box>

      <ModalCreateProduct
        isLoading={isLoadingCreate}
        isAdmin={isAdmin}
        listUser={listUser}
        isOpen={isOpenModalCreateProduct}
        idCustomer={id}
        itemUpdate={{}}
        onClose={handleCloseModalCreateProduct}
        onSubmit={handleCreateProduct}
      />
    </>
  );
};

export default memo(ListProductContent);
