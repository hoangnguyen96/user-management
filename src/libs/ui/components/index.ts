// Common
export * from "./Common";

// Components
export { default as HeadingContent } from "./HeadingContent";
export { default as LazyFallbackContent } from "./LazyFallbackContent";
export { default as Logo } from "./Logo";
export { default as Footer } from "./Footer";
export { default as FormController } from "./FormController";
export { default as ListCustomersContent } from "./ListCustomersContent";
export { default as ListProductContent } from "./ListProductContent";
export { default as NavigationBar } from "./NavigationBar";
export { default as Temporary } from "./Temporary";
export { default as TitleContent } from "./TitleContent";
export { default as MenuBar } from "./MenuBar";
export { default as ErrorBoundary } from "./ErrorBoundary";
export { default as ModalCreateUserSuccess } from "./Modal/ModalCreateUserSuccess";
