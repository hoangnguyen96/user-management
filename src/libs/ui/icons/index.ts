export * from "./HelpIcon";
export * from "./IncomeIcon";
export * from "./KeyIcon";
export * from "./Members";
export * from "./ProductIcon";
export * from "./PromoteIcon";
export * from "./Screen";
export * from "./SearchIcon";
export * from "./TotalCustomers";
export * from "./UserIcon";
