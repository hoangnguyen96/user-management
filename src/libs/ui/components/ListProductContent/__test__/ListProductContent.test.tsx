import { WrapperRouter, renderWithQueryClient } from "@app/ui/test-utils";
import { fireEvent, waitFor } from "@testing-library/react";
import { useMatch } from "react-router-dom";

// Api
import * as api from "@app/api";

// Utils
import * as utils from "@app/utils";

// Services
import { HttpClient } from "@app/services";

// Component
import ListProductContent from "..";
import {
  LIST_CONTENT_PRODUCT,
  LIST_CONTENT_PRODUCT_PAGE1,
  LIST_CONTENT_PRODUCT_PAGE2,
} from "../__mocks__/mock-data";

const mockCustomer = [
  {
    fullName: "Jane Cooper",
    role: "admin",
    company: "Microsoft",
    phoneNumber: "2255550118",
    email: "jane@microsoft.com",
    country: "United States",
    status: false,
    username: "admin",
    password: "123456",
    id: "1",
  },
  {
    fullName: "Floyd Miles",
    role: "consumer",
    company: "Yahoo",
    phoneNumber: "2055550100",
    email: "floyd@yahoo.com",
    country: "Kiribati",
    status: true,
    username: "consumer1",
    password: "123456",
    id: "2",
  },
];

const productListData = [
  {
    name: "Hamburger",
    price: 32,
    quantity: 69,
    status: false,
    image: "/src/libs/ui/images/img2.jpg",
    code: "POD2",
    id: "2",
    userId: "2",
  },
];

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useMatch: jest.fn(),
}));

jest.mock("@app/utils", () => ({
  ...jest.requireActual("@app/utils"),
  isEnableSubmitButton: jest.fn(),
  dividePaginationProduct: jest.fn(() => [
    LIST_CONTENT_PRODUCT_PAGE1,
    LIST_CONTENT_PRODUCT_PAGE2,
  ]),
}));

jest.mock("@app/api", () => ({
  useGetListProduct: jest.fn(() => ({
    isLoading: false,
    refetch: jest.fn(),
    data: LIST_CONTENT_PRODUCT,
    errorMessage: "",
  })),
  useGetListProductOfUser: jest.fn(() => ({
    isLoading: false,
    data: productListData,
    refetch: jest.fn(),
    errorMessage: "",
  })),
  useGetListCustomers: jest.fn(() => ({
    data: mockCustomer,
  })),
  useCreateProduct: jest.fn(() => ({
    isPending: false,
    mutate: jest.fn(),
  })),
  useGetProductById: jest.fn(() => ({
    refetch: jest.fn(),
    data: {},
    errorMessage: null,
  })),
  useUpdateProduct: jest.fn(() => ({
    isPending: false,
    mutate: jest.fn(),
  })),
  useDeleteProduct: jest.fn(() => ({
    isPending: false,
    mutate: jest.fn(),
  })),
}));

const spyPost = jest.spyOn(HttpClient, "post");

describe("ListProductContent component", () => {
  const mockUseMatch = useMatch as jest.MockedFunction<typeof useMatch>;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should render snapshot correctly", () => {
    expect(
      renderWithQueryClient(
        <WrapperRouter>
          <ListProductContent isAdmin={true} id="1" />
        </WrapperRouter>
      )
    ).toMatchSnapshot();
  });

  it("Should handle text input change correctly", async () => {
    const { getByPlaceholderText } = renderWithQueryClient(
      <WrapperRouter>
        <ListProductContent isAdmin={true} id="1" />
      </WrapperRouter>
    );

    const input = getByPlaceholderText("Searchâ€¦") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "John" } });

    await waitFor(() => {
      expect(input.value).toBe("John");
    });
  });

  it("Should handle change pagination correctly", () => {
    const { getByRole } = renderWithQueryClient(
      <WrapperRouter>
        <ListProductContent isAdmin={true} id="1" />
      </WrapperRouter>
    );
    const paginationButton = getByRole("button", { name: "Go to page 2" });
    fireEvent.click(paginationButton);

    expect(paginationButton).toBeInTheDocument();
  });

  it("Should handle close modal add product", async () => {
    mockUseMatch.mockReturnValue({
      params: {},
      pathname: "/product",
      pathnameBase: "/product",
      pattern: { path: "/product" },
    });
    const { getByTestId, getByText } = renderWithQueryClient(
      <WrapperRouter>
        <ListProductContent isAdmin={false} id="1" />
      </WrapperRouter>
    );

    fireEvent.click(getByTestId("modal-add-product"));

    await waitFor(() => {
      expect(getByText(/ADD PRODUCT/i)).toBeInTheDocument();
    });

    // Close modal
    fireEvent.click(getByTestId("close-modal-button"));
  });

  it("Should handle sorted for selection list", async () => {
    const { getByRole } = renderWithQueryClient(
      <WrapperRouter>
        <ListProductContent isAdmin={false} id="1" />
      </WrapperRouter>
    );

    const combobox = getByRole("combobox", {
      name: /short by:/i,
    });
    fireEvent.mouseDown(combobox);

    const newOption = getByRole("option", {
      name: /price/i,
    });
    fireEvent.click(newOption);

    expect(newOption).toBeInTheDocument();
  });

  it("Should handle create product successful", async () => {
    jest.spyOn(utils, "isEnableSubmitButton").mockReturnValue(true);
    const mockCreateProductSuccess = jest
      .fn()
      .mockImplementation((_, { onSuccess }) => {
        onSuccess({
          code: "POD013",
          name: "Test Name",
          quantity: 12,
          image: "https://i.ibb.co/vkbZPGr/img10.jpg",
          price: 3,
          userId: "6",
        });
      });
    (api.useCreateProduct as jest.Mock).mockReturnValue({
      isisPending: false,
      mutate: mockCreateProductSuccess,
    });
    spyPost.mockImplementationOnce((): Promise<unknown> => Promise.resolve());
    mockUseMatch.mockReturnValue({
      params: {},
      pathname: "/product",
      pathnameBase: "/product",
      pattern: { path: "/product" },
    });

    const { getByTestId, getByText, getByRole } = renderWithQueryClient(
      <WrapperRouter>
        <ListProductContent isAdmin={true} id="1" />
      </WrapperRouter>
    );

    fireEvent.click(getByTestId("modal-add-product"));

    await waitFor(() => {
      expect(getByText(/ADD PRODUCT/i)).toBeInTheDocument();
    });

    const inputNameProduct = getByTestId("name-product").querySelector(
      "input"
    ) as HTMLInputElement;
    const inputPriceProduct = getByTestId("price-product").querySelector(
      "input"
    ) as HTMLInputElement;
    const inputQuantityProduct = getByTestId("quantity-product").querySelector(
      "input"
    ) as HTMLInputElement;

    // Ensure the select element is rendered
    await waitFor(() => {
      expect(getByTestId("select-user")).toBeInTheDocument();
    });

    // Use getByTestId directly to get the select element
    const selectName = getByRole("combobox", {
      name: /without label/i,
    });

    expect(selectName).not.toBeNull();

    expect(inputNameProduct).toHaveProperty("value");
    expect(inputPriceProduct).toHaveProperty("value");
    expect(inputQuantityProduct).toHaveProperty("value");

    fireEvent.change(inputNameProduct, {
      target: { value: "Banana" },
    });
    fireEvent.change(inputPriceProduct, {
      target: { value: 1 },
    });
    fireEvent.change(inputQuantityProduct, {
      target: { value: 2 },
    });

    // Open the dropdown
    fireEvent.mouseDown(selectName);
    const newOption = getByRole("option", {
      name: /floyd miles/i,
    });
    fireEvent.click(newOption);

    // Submit the form
    fireEvent.click(getByTestId("button-product"));

    await waitFor(() => {
      expect(getByText(/ADD PRODUCT/i)).toBeInTheDocument();
    });
  });
});
