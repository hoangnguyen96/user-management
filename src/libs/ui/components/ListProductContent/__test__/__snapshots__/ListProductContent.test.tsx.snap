// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListProductContent component Should render snapshot correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1dachdb"
      >
        <div
          class="MuiStack-root css-133lg72-MuiStack-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-15gon46-MuiTypography-root"
            >
              All Product
            </h4>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-1ge8cc0-MuiTypography-root"
            >
              Active Product
            </h6>
          </div>
          <div
            class="MuiStack-root css-qoybat-MuiStack-root"
          >
            <div
              class="MuiBox-root css-79elbk"
            >
              <div
                class="MuiStack-root css-1ibunwk-MuiStack-root"
              >
                <svg
                  fill="none"
                  height="24px"
                  viewBox="0 0 24 24"
                  width="24px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11 19C15.4183 19 19 15.4183 19 11C19 6.58172 15.4183 3 11 3C6.58172 3 3 6.58172 3 11C3 15.4183 6.58172 19 11 19Z"
                    stroke="#7e7e7e"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                  <path
                    d="M21 21L16.65 16.65"
                    stroke="#7e7e7e"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-sfs0yr-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-label="search"
                  class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  placeholder="Search…"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root css-2iaed4-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-wxw8j3-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                id="demo-simple-select-label"
              >
                Short by:
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-cb5d54-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                data-testid="change-selection"
              >
                <div
                  aria-controls=":r0:"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="demo-simple-select-label demo-simple-select"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-zbhh1g-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="demo-simple-select"
                  role="combobox"
                  tabindex="0"
                >
                  -select-
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value="default"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-10d36jd-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="KeyboardArrowDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-wxj3k5-MuiStack-root"
        >
          <div
            class="MuiStack-root css-156dzxc-MuiStack-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-6r59v1-MuiTypography-root"
            >
              Product Code
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-jesh3t-MuiTypography-root"
            >
              Name
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-6r59v1-MuiTypography-root"
            >
              Price
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-16972la-MuiTypography-root"
            >
              Quantity
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-184m8mr-MuiTypography-root"
            >
              Image
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-w9byz0-MuiTypography-root"
            >
              Status
            </h6>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-yk10cl-MuiDivider-root"
          />
          <span
            class="MuiTypography-root MuiTypography-caption css-en5znm-MuiTypography-root"
          >
            Data Null
          </span>
          <div
            class="MuiStack-root css-bmu17z-MuiStack-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-1jwn8bo-MuiTypography-root"
            >
              Showing data 
              1
               to
               
              0
               of
               
              0
               entries
            </h6>
            <nav
              aria-label="pagination navigation"
              class="MuiPagination-root MuiPagination-outlined css-1r9e4h5-MuiPagination-root"
              data-testid="change-pagination"
            >
              <ul
                class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
              >
                <li>
                  <button
                    aria-label="Go to previous page"
                    class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-previousNext css-ahvmcb-MuiButtonBase-root-MuiPaginationItem-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                      data-testid="NavigateBeforeIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                      />
                    </svg>
                  </button>
                </li>
                <li>
                  <button
                    aria-label="Go to next page"
                    class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-previousNext css-ahvmcb-MuiButtonBase-root-MuiPaginationItem-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                      data-testid="NavigateNextIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                  </button>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1dachdb"
    >
      <div
        class="MuiStack-root css-133lg72-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-15gon46-MuiTypography-root"
          >
            All Product
          </h4>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-1ge8cc0-MuiTypography-root"
          >
            Active Product
          </h6>
        </div>
        <div
          class="MuiStack-root css-qoybat-MuiStack-root"
        >
          <div
            class="MuiBox-root css-79elbk"
          >
            <div
              class="MuiStack-root css-1ibunwk-MuiStack-root"
            >
              <svg
                fill="none"
                height="24px"
                viewBox="0 0 24 24"
                width="24px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 19C15.4183 19 19 15.4183 19 11C19 6.58172 15.4183 3 11 3C6.58172 3 3 6.58172 3 11C3 15.4183 6.58172 19 11 19Z"
                  stroke="#7e7e7e"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
                  d="M21 21L16.65 16.65"
                  stroke="#7e7e7e"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-sfs0yr-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-label="search"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                placeholder="Search…"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root css-2iaed4-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-wxw8j3-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              id="demo-simple-select-label"
            >
              Short by:
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-cb5d54-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              data-testid="change-selection"
            >
              <div
                aria-controls=":r0:"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="demo-simple-select-label demo-simple-select"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-zbhh1g-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="demo-simple-select"
                role="combobox"
                tabindex="0"
              >
                -select-
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="default"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-10d36jd-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiStack-root css-wxj3k5-MuiStack-root"
      >
        <div
          class="MuiStack-root css-156dzxc-MuiStack-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-6r59v1-MuiTypography-root"
          >
            Product Code
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-jesh3t-MuiTypography-root"
          >
            Name
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-6r59v1-MuiTypography-root"
          >
            Price
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-16972la-MuiTypography-root"
          >
            Quantity
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-184m8mr-MuiTypography-root"
          >
            Image
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-w9byz0-MuiTypography-root"
          >
            Status
          </h6>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-yk10cl-MuiDivider-root"
        />
        <span
          class="MuiTypography-root MuiTypography-caption css-en5znm-MuiTypography-root"
        >
          Data Null
        </span>
        <div
          class="MuiStack-root css-bmu17z-MuiStack-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-1jwn8bo-MuiTypography-root"
          >
            Showing data 
            1
             to
             
            0
             of
             
            0
             entries
          </h6>
          <nav
            aria-label="pagination navigation"
            class="MuiPagination-root MuiPagination-outlined css-1r9e4h5-MuiPagination-root"
            data-testid="change-pagination"
          >
            <ul
              class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
            >
              <li>
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-previousNext css-ahvmcb-MuiButtonBase-root-MuiPaginationItem-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateBeforeIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                    />
                  </svg>
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-previousNext css-ahvmcb-MuiButtonBase-root-MuiPaginationItem-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateNextIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                    />
                  </svg>
                </button>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
